services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker_core.py

```diff
@@ -1,135 +1,91 @@
-import datetime
+import asyncio
 import json
 import logging
 from typing import Any
 
 import arrow
 from fastapi import FastAPI
-from models_library.users import UserID
 from prometheus_api_client import PrometheusConnect
 from simcore_service_resource_usage_tracker.modules.prometheus import (
     get_prometheus_api_client,
 )
 
-# This script assumes everywhere that the minimum granularity of the data is 1 minute
-# Setting it smaller is unreasonable at least for prometheus, as the scraping interval is apprx. eq. to 1 h
-
+from .models.resource_tracker_container import ContainerResourceUsage
+from .modules.db.repositories.resource_tracker import ResourceTrackerRepository
 
 _logger = logging.getLogger(__name__)
 
 
-def _assure_dict_entry_exists(
-    metric_data, max_values_per_docker_id, image, userid
-) -> None:
-    for metric in metric_data:
-        current_id = metric["metric"]["id"]
-        if current_id not in max_values_per_docker_id.keys():
-            max_values_per_docker_id[current_id] = {
-                "container_uuid": metric["metric"]["container_label_uuid"],
-                "cpu_seconds": 0,
-                "uptime_minutes": 0,
-                "nano_cpu_limits": 0,
-                "egress_bytes": 0,
-                "image": image,
-                "user_id": userid,
-            }
+async def _prometheus_client_custom_query(
+    prometheus_client: PrometheusConnect, promql_cpu_query: str
+) -> list[dict]:
+    _logger.info("Querying prometheus with: %s", promql_cpu_query)
+    data: list[dict] = await asyncio.get_event_loop().run_in_executor(
+        None, prometheus_client.custom_query(promql_cpu_query)
+    )
+    return data
 
 
-async def _evaluate_service_resource_usage(
+async def _scrape_and_upload_container_resource_usage(
     prometheus_client: PrometheusConnect,
-    start_time: datetime.datetime,
-    stop_time: datetime.datetime,
-    user_id: UserID,
-    uuid: str = ".*",
-    image: str = "registry.osparc.io/simcore/services/dynamic/jupyter-smash:3.0.9",
-) -> dict[str, Any]:
-    max_values_per_docker_id: dict[str, Any] = {}
-    time_delta = stop_time - start_time
-    minutes = round(time_delta.total_seconds() / 60)
+    resource_tracker_repo: ResourceTrackerRepository,
+    image_regex: str = "registry.osparc.io/simcore/services/dynamic/jupyter-smash:.*",
```

Comment:
I would suggest to you to already now make this image_regex fully customization by adding at is a env-var setting, akin to `PROMETHEUS_URL`. We will for sure not only monitor jupyter-smash ;)

Reply:
For now, I would like to leave there just jupyter-smash (so we can quickly get to the production and get some feedback). In additional iterations, we need to think a list of regex expressions which will catch all the containers that we want to scrape (we need to be sure that in the output of GraphQL there is not much data (as GraphQL doesn't support pagination) therefore probably we will run more this tasks that will scrape different containers.

Reply:
With GraphQL I assume you mean PromQL?

In my mind I read what you propose (handling multiple regex, i.e. a list, as building a monolith in disguise. I think it would come much more natural to the microservice/container ecosystem if we had one resrouce tracking microservice instance per docker image that we want to monitor. The autoscaling microservice was designed with this design philosophy and being able to run multiple autoscaling instances for different oSparc dynamic services or different conditions has been hugely beneficial in the past to move quickly.

If you want, maybe comment or chat with me ;) Anyway, I personally would feel like it might not be wise to add monitoring for a list of services in this microservice, but of course my gut feeling might be wrong. I also feel like multiple slim microservices are easier to operate than one big one.

Reply:
Added a comment to the top after in-person chat